ICP Dev Docs Restructure, Overhaul & Audit (#3970)
dfinity/portal: docs/building-apps/best-practices/security/data-storage.mdx

Please note some of these libraries may be partially unfinished.
:::

- See [current limitations of the Internet Computer](https://wiki.internetcomputer.org/wiki/Current_limitations_of_the_Internet_Computer), sections "Long running upgrades" and "\[de\]serializer requiring additional Wasm memory.".

- For example, [Internet Identity](https://github.com/dfinity/internet-identity) uses stable memory directly to store user data.

## Consider encrypting sensitive data on canisters

### Security concern

By default, canisters provide integrity but not confidentiality. Data stored on canisters can be read by nodes/replicas.

### Recommendation

- Consider end-to-end encrypting any private or personal data (e.g., a user’s personal or private information) on canisters.

- The example dapp [encrypted notes](https://github.com/dfinity/examples/tree/master/motoko/encrypted-notes-dapp) illustrates how end-to-end encryption can be done.

## Create backups

### Security concern

> A canister could be rendered unusable and impossible to upgrade again, e.g., due to reasons such as:

Reviewer: ```suggestion
A canister could be rendered unusable and impossible to upgrade. For example, due to one of the following reasons:
```
